input/drivers/dinput.c

```diff
@@ -181,6 +182,41 @@ static void *dinput_init(const char *joypad_driver)
    return di;
 }
 
+static void dinput_keyboard_mods(struct dinput_input *di, int mod)
+{
+   switch (mod)
+   {
+      case RETROKMOD_SHIFT:
+         if ( ((GetAsyncKeyState(VK_LSHIFT) >> 1) && !di->shift_l) ||
+             (!(GetAsyncKeyState(VK_LSHIFT) >> 1) && di->shift_l))
+         {
+            input_keyboard_event((GetAsyncKeyState(VK_LSHIFT) >> 1), RETROK_LSHIFT,
+                  0, RETROKMOD_SHIFT, RETRO_DEVICE_KEYBOARD);
+            di->shift_l = !di->shift_l;
+         }
+
+         if ( ((GetAsyncKeyState(VK_RSHIFT) >> 1) && !di->shift_r) ||
+             (!(GetAsyncKeyState(VK_RSHIFT) >> 1) && di->shift_r))
+         {
+            input_keyboard_event((GetAsyncKeyState(VK_RSHIFT) >> 1), RETROK_RSHIFT,
+                  0, RETROKMOD_SHIFT, RETRO_DEVICE_KEYBOARD);
+            di->shift_r = !di->shift_r;
+         }
+         break;
+
+      case RETROKMOD_ALT:
+         if ((GetAsyncKeyState(VK_LMENU) >> 1) && !di->alt_l)
+            di->alt_l = !di->alt_l;
+         else if (!(GetAsyncKeyState(VK_LMENU) >> 1) && di->alt_l)
```

Comment:
Is it possible you can cache the results of GetAsyncKeyState(VK_LMENU) here once instead of having to call it twice in case the first conditional fails? Of course, if you do this, make sure that the code is still C89-compliant by declaring the temporary variable either at the start of a code block or the start of the function.